active/003-seq-feature.md

```diff
@@ -0,0 +1,126 @@
+---
+Feature name: interval-metadata
+Start date: <2015-12-18>
+Pull request: 1
+Authors: ["@RNAer", "@mortonjt"]
+Contributors: ["@RNAer", "@mortonjt", "@ebolyen", "@jairideout", "@wasade", "@gregcaporaso", "@rob-knight"]
+---
+
+# Summary
+
+Introduce an `interval_metadata` object to allow for the storage, modification, and querying of features covering a region of a biological sequence. A feature is defined as a sub-region of a seuqence that is a functional entity, e.g., a gene, a riboswitch, an exon, etc.
+
+# Motivation
+
+The current implementation of `positional_metadata` in the `Sequence` class takes > 1TB memory to read in a E. coli genome with 260K features. And the `pandas` sparse data frame can alleviate the mem usage, but it is still too memory hogging. The discussion is mainly in this [thread](https://github.com/biocore/scikit-bio/issues/1159).  Storing intervals instead of `positional_metadata` would substantially save on memory.
+
+Additional benefits of having a `interval_metadata` objects include the ability to rapidly query features by coordinates.  Interval trees allow for both querying by position and ranges to retrieve all overlapping intervals.  This can be particularly helpful when dealing with layers of features (i.e. transcripts made up of exons, or operons made up of genes).
+
+# Detailed design
+
+We propose 2 new public data objects: `BoundFeature` and `IntervalMetadata`. `BoundFeature` stores all the attributes of a feature. `IntervalMetadata` stores all the `BoundFeatures` objects of a sequence and add it as `interval_metadata` attribute (as similar to `positional_metadata`) in `Sequence` (and its child classes).
+
+## `BoundFeature` object
```

Comment:
Since features can be bound or unbound, it may be valuable to have two public types, `Feature` and `BoundFeature`. Users use `Feature` to create a new unbound feature, and when `Features` are added to `IntervalMetadata` they become `BoundFeatures`, which users can retrieve and mutate as necessary.

If we don't make a distinction between bound and unbound features, the following example isn't straightforward to reason about:

```python
feat = BoundFeature(intervals=[1, 2], foo='bar') # `feat` is currently unbound, the type `BoundFeature` is a confusing name
im1 = IntervalMetadata()
im2 = IntervalMetadata()
im1.add(feat) # `feat` is now bound to `im1`
im2.add(feat) # `feat` is now bound to `im2`
feat.intervals = [3, 4] # `im2` is updated because it has the latest binding, `im1` is not
feat.foo = 'baz' # is `im1`, `im2`, or both `im1` and `im2` updated?
```

If there were a distinction, it becomes much easier to reason about:

```python
feat = Feature(intervals=[1, 2], foo='bar') # type `Feature` is *never* bound
im1 = IntervalMetadata()
im2 = IntervalMetadata()
im1.add(feat) # `feat` is stored as a `BoundFeature` on `im1`
im2.add(feat) # `feat` is stored as a `BoundFeature` on `im2`
feat.intervals = [3, 4] # `im1` and `im2` are not updated because the unbound feature is mutated
feat.foo = 'baz' # `im1` and `im2` are not updated because the unbound feature is mutated
bound_feat1 = next(im1.query(foo='bar')) # `bound_feat1` is a `BoundFeature` on `im1`
bound_feat2 = next(im2.query(foo='bar')) # `bound_feat2` is a `BoundFeature` on `im2`
bound_feat1.intervals = [3, 4] # only `im1` is updated
bound_feat2.intervals = [3, 4] # only `im2` is updated
bound_feat1.foo = 'baz' # only `im1` is updated
bound_feat2.foo = 'baz' # only `im2` is updated
```